pkg/api/validation/events.go

```diff
@@ -25,20 +31,55 @@ import (
 // ValidateEvent makes sure that the event makes sense.
 func ValidateEvent(event *api.Event) field.ErrorList {
 	allErrs := field.ErrorList{}
-	// There is no namespace required for node.
+	// There is no namespace required for root-scoped kind, for example, node.
 	// However, older client code accidentally sets event.Namespace
 	// to api.NamespaceDefault, so we accept that too, but "" is preferred.
-	if event.InvolvedObject.Kind == "Node" &&
-		event.Namespace != api.NamespaceDefault &&
-		event.Namespace != "" {
-		allErrs = append(allErrs, field.Invalid(field.NewPath("involvedObject", "namespace"), event.InvolvedObject.Namespace, "not allowed for node"))
-	}
-	if event.InvolvedObject.Kind != "Node" &&
-		event.Namespace != event.InvolvedObject.Namespace {
-		allErrs = append(allErrs, field.Invalid(field.NewPath("involvedObject", "namespace"), event.InvolvedObject.Namespace, "does not match involvedObject"))
+	// Todo: Events may reference 3rd party object, and we can't check whether the object is namespaced.
+	// Suppose them are namespaced. Do check if we can get the piece of information.
+	// This should apply to all groups served by this apiserver.
+	group := apiutil.GetGroup(event.InvolvedObject.APIVersion)
+	version := apiutil.GetVersion(event.InvolvedObject.APIVersion)
+	if (group == "" || group == "extensions") && (version != "extensions" && version != "") {
+		namespacedKindFlag, err := isNamespacedKind(event.InvolvedObject.Kind, event.InvolvedObject.APIVersion)
+		if err != nil {
+			allErrs = append(allErrs, field.Invalid(field.NewPath("involvedObject", "kind"), event.InvolvedObject.Kind, fmt.Sprintf("couldn't check whether namespace is allowed: %s", err)))
+		} else {
+			if !namespacedKindFlag &&
+				event.Namespace != api.NamespaceDefault &&
```

Comment:
To ```Why do our events come in for namespace="default" when the involved object is namespace=""?```, because older client code accidentally sets event.Namespace to api.NamespaceDefault, we should accept this (the comment on the codes).